qsk-macros/src/parse.rs

```diff
@@ -0,0 +1,424 @@
+                    //                                                        ^
+                    TokenTree::Punct(punct) => {
+                        if punct.as_char() != ',' {
+                            return Err(cursor.error("unexpected punctuation"))
+                        }
+                    },
+                    // if we find any additional token trees then something is wrong.
+                    _ => return Err(cursor.error(format!("unexpected token tree: {:?}", name))),
+                }
+            }
+
+            return Ok((ControlCode::Function(key_func), rest));
+        })
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct KeyMaps {
+    pub(crate) lhs: Key,
+    pub(crate) rhs: ControlCode,
+}
+
+impl Parse for KeyMaps {
+    fn parse(stream: ParseStream) -> Result<Self> {
+        let lhs = Key(stream.parse()?);
+        stream.parse::<Token![->]>()?; // discard operator for now
+        let rhs = stream.parse()?;
+        Ok(KeyMaps{
+            lhs,
+            rhs,
+        })
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct LayerBody {
+    pub(crate) maps: Punctuated<KeyMaps, Token![,]>,
+}
+
+impl LayerBody {
+    pub fn iter(&self) -> impl Iterator<Item = &KeyMaps> {
+        self.maps.iter()
+    }
+}
+
+impl Parse for LayerBody {
+    fn parse(stream: ParseStream) -> Result<Self> {
+        let content;
+        braced!(content in stream);
+        Ok(LayerBody{
+            maps: content.parse_terminated(KeyMaps::parse)?,
+        })
+    }
+}
+
+pub struct LayerOpts {
+    pub(crate) opts: Punctuated<Ident, Token![,]>,
+}
+
+impl Parse for LayerOpts {
+    fn parse(stream: ParseStream) -> Result<Self> {
+        let content;
+        bracketed!(content in stream);
+        Ok(LayerOpts{
+            opts: content.parse_terminated(Ident::parse)?,
+        })
+    }
+}
+
+pub struct Layer {
+    pub(crate) name: Ident,
+    pub(crate) opts: Option<LayerOpts>,
+    pub(crate) body: LayerBody,
+}
+
+impl Parse for Layer {
+    fn parse(stream: ParseStream) -> Result<Self> {
+        let name = stream.parse()?;
+        let mut opts: Option<LayerOpts> = None;
+        if let Ok(o) = stream.parse() {
+            opts = Some(o);
+        }
+        stream.parse::<Token![:]>()?;
+        Ok(Layer {
+            name,
+            opts,
+            body: stream.parse()?,
+        })
+    }
+}
+
+pub struct Ast {
+    pub(crate) layers: Punctuated<Layer, Token![,]>,
+}
+
+impl Ast {
+    pub fn iter(&self) -> impl Iterator<Item = &Layer> {
+        self.layers.iter()
+    }
+}
+
+impl Parse for Ast {
+    fn parse(stream: ParseStream) -> Result<Self> {
+        let layers = stream.parse_terminated(Layer::parse)?;
+        Ok(Ast {
+            layers,
+        })
+    }
+}
+
+pub fn parse(ts: TokenStream) -> Ast {
+    match parse2::<Ast>(ts) {
+        Ok(ast) => ast,
+        Err(e) => {
+            abort!(e.span(), e)
```

Comment:
TODO: don't abort here, only abort in the procedural macro itself since `proc_macro_error::abort` can strictly only be used in the actual function that takes a token stream and outputs a token stream. calling it here reduces testability because the `proc_macro_error::abort` call panics wildly.